src/main/java/de/tum/cit/aet/artemis/core/service/CourseService.java

```diff
@@ -652,6 +652,19 @@ public List<Course> getAllCoursesForManagementOverview(boolean onlyActive) {
         return courseRepository.findAllCoursesByManagementGroupNames(userGroups);
     }
 
+    /**
+     * Fetches all courses, filters out courses that the user does not have
+     * access to based on their role, and excludes any courses that do not
+     * belong to a specific semester.
+     *
+     * @return A list of courses for the course archive
+     */
+    public List<Course> getAllCoursesForCourseArchive() {
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        List<Course> courses = courseRepository.findAll();
+        return courses.stream().filter(course -> authCheckService.isAtLeastStudentInCourse(course, user) && course.getSemester() != null).toList();
+    }
```

Comment:
is it possible to do this on a database level?
otherwise we keep all courses in the memory which could be a basic query.

Reply:
Makes sense to me, I will handle it on a database level 👍